Refresh the Test Replay page (#5985)
cypress-io/cypress-documentation: docs/cloud/features/test-replay.mdx

:::

Test Replay in Cypress Cloud allows developers to accurately and efficiently debug failed and flaky [continuous integration](/app/continuous-integration/overview) test runs. It captures every test run detail, enabling developers to replay it and inspect the DOM, network requests, console logs, JavaScript errors, and element rendering as they happened during `cypress run`.

<DocsVideo
  src="https://youtube.com/embed/vFLShoCM8pA"
  title="Test Replay Product Demo"
/>

## Key Benefits

Test failures in CI have historically been captured through screenshots, videos, and stack trace outputs,
but these artifacts provide limited information. Developers often need more context to understand the
root cause of failures, requiring them to reproduce the issue locally, which may not always
be possible due to environmental or network conditions.
This knowledge gap between test failures and debugging capabilities leads to increased debugging

> time and potentially deploying unresolved bugs to production.

Reviewer: The transition into the starred list seems a little abrupt. Maybe an intro to the list of some kind?

```suggestion
time and potentially deploying unresolved bugs to production. Test Replay solves these problems by providing the following features.
```

or something

<CloudFreePlan />

:::

### 1. Record Test Runs

All you need to do is [record test runs](/cloud/get-started/setup) and log in to your Cypress Cloud
account to leverage Test Replay's debugging capabilities. No additional configuration
or changes to your existing tests are required.

### 2. Access Test Replay

Accessing the replay of your test is as simple as reviewing any other test artifact in Cypress Cloud.

Hover on a [test for review](/cloud/features/recorded-runs#Tests-for-Review) under a run's **Overview**.

<DocsImage
  src="/img/cloud/features/test-replay/replay_row_magnified.jpg"
  alt="Run overview row"
/>

Or navigate to the [test detail sidebar](/cloud/features/recorded-runs#Test-detail-sidebar) header or above each attempt in the errors panel. The header option will replay the first failing test while the option above each attempt will replay the affiliated test.

<DocsImage
  src="/img/cloud/features/test-replay/replay_sidebar_magnified.jpg"
  alt="Run sidebar"
  width="60%"
/>

### 3. Debug With Test Replay

Both options above display the Test Replay interface, which we'll unpack from top to bottom in more detail below.

#### Run header

> The left side of the header displays the Cypress Cloud test run ID number (#), with the Git branch name along with the spec, platform and browser information. Hovering on this exposes additional test meta data.

Reviewer: ```suggestion
The left side of the header displays the Cypress Cloud test run ID number (#), with the Git branch name along with the spec, platform and browser information. Hovering on this exposes additional test metadata.
```

Cypress captures a wide range of data during end-to-end and component test runs.
This data is then uploaded and stored in Cypress Cloud and replayed for debugging through Test Replay.

### What is captured

- Test commands
- Network requests
- Console logs
- JavaScript errors
- Element rendering
- DOM mutations
- CSS styles
- SVG elements
- Iframes
- Shadow DOM
- Canvas elements (v13.5.0+, please contact us at support@cypress.io to enable capturing canvas elements)

### What is not captured

Currently Test Replay does not support the following:

- Tests recorded in Cypress App below v13.
- WebKit browser (Safari) ([see issue](https://github.com/cypress-io/cypress/issues/28895))
- Firefox browser ([see issue](https://github.com/cypress-io/cypress/issues/28894))
- Video ([see issue](https://github.com/cypress-io/cypress/issues/29903))
- Audio ([see issue](https://github.com/cypress-io/cypress/issues/29905))
- Objects that do not have `type="image/svg+xml"` ([see issue](https://github.com/cypress-io/cypress/issues/29904))
- Shadow DOM with `slotAssignment="manual"` ([see issue](https://github.com/cypress-io/cypress/issues/29107))

> - Shadow DOM canvas ([see issue](https://github.com/cypress-io/cypress/issues/30438))

Reviewer: ```suggestion
- Shadow DOM canvases ([see issue](https://github.com/cypress-io/cypress/issues/30438))
```

Author: Updated here to be more specific on plurality and elements vs other things [`56e77c8` (#5985)](https://github.com/cypress-io/cypress-documentation/pull/5985/commits/56e77c8547c7ac7c4ac3752e6e44a73170bfc15c)

- Server sent events ([see issue](https://github.com/cypress-io/cypress/issues/29908))
- Web sockets ([see issue](https://github.com/cypress-io/cypress/issues/29907))
- localStorage & sessionStorage ([see issue](https://github.com/cypress-io/cypress/issues/29909))
- Cookies ([see issue](https://github.com/cypress-io/cypress/issues/29910))

> - Console props of Cypress commands outside of `cy.request()` ([see issue](https://github.com/cypress-io/cypress/issues/29911))

Reviewer: Do we want to call out that we do not capture network traffic of `cy.request()` either?

Author: Addressed here [`7ec343a` (#5985)](https://github.com/cypress-io/cypress-documentation/pull/5985/commits/7ec343a678e0c8096d598249fe0f73823421f37a) I created a public issue to track this also.

### Performance

#### Recording

Capturing this additional data can use additional resources while recording, but you can
expect an even performance exchange with Test Replay versus video recording.
There is more activity to capture Test Replay data but without the overhead of recording,
compressing, and storing video assets.

:::info

> **Tip:** We recommend disabling video recording when recording with

Reviewer 2: is it clear that video is disabled by default?

Author: There are so many customers I've seen that have it on, I do think it needs to be called out very explicitly.
